rasterio/features.py

```diff
@@ -268,7 +268,7 @@ def rasterize(
 
     # The output data type is determined by the output array or dtype
     # parameter.
-    dtype = out.dtype.name if out is not None else dtype
+    dtype = out.dtype.name if out is not None else (np.dtype(dtype).name if dtype is not None else dtype)
```

Comment:
The line is getting a bit long. Thoughts about adding another line?
```suggestion
    dtype = np.dtype(dtype).name if dtype is not None else dtype
    dtype = out.dtype.name if out is not None else dtype
```

Reply:
This might be more clear:
```python
if out is not None:
    dtype = out.dtype.name
elif dtype is not None:
    dtype = np.dtype(dtype).name
else:
    dtype = dtype
```

Reply:
I like the solution proposed by @groutr

Reply:
Could probably remove the final `else`

Reply:
Thanks @groutr !